### ISSUE TITLE
2-Phase Borrows

### ISSUE THREAD

**Issue by arielb1** [nikomatsakis/nll-rfc#37]
The NLL RFC does not include any description of 2-phase borrows, which are required in order to enable nested method calls aka rust-lang/rfcs#2025:

```Rust
x.push(x.len());
```

Or the MIR equivalent
```
tmp0 = &'a mut vec;
tmp1 = &'b vec;
tmp2 = Vec::len(tmp1);
Vec::push(tmp0, tmp2);
```

**Comment by nikomatsakis** [comment]
Yeah, I intentionally left them out because I kind of ran out of time and it seemed fairly orthogonal to me. But it would be good to include. I was intending to simply "port" the approach described in rust-lang/rfcs#2025, at least for now.
